### ISSUE TITLE
Avoid changing the prototype of native Promise

### ISSUE THREAD

**Issue by danieljuhl** [SparkPost/node-sparkpost#199]
It is a very bad practice to change the global prototype for native/builtins.

If I require the sparkpost module, my Promise won't be a real native Promise anymore, as you are adding `.asCallback()` method in `lib/Promise.js`.

Could you please avoid doing this? If you needs such functionality internally, you can easily create a simple wrapper function instead, and just import and use that function.

```
function asCallback(promise, cb) {
  if (typeof cb !== 'function') {
    cb = noop;
  }

  return promise.then((result) => {
    cb(null, result);
    return promise;
  }).catch((err) => {
    cb(err);
    return promise;
  });
};

function noop() {}
```

And then simply use it as eg. `return asCallback(client.get(options), callback);` instead of `client.get(options).asCallback(callback);`

If the above is accepted, `withCallback` is probably a better name than `asCallback`

**Comment by danieljuhl** [comment]
The reason I stumbled on this was the following error trace in my monitoring system:

```
TypeError: this.request(...).asCallback is not a function
at SparkPost.post (.../node_modules/sparkpost/lib/sparkpost.js:150:32)
at Object.send (.../node_modules/sparkpost/lib/transmissions.js:81:21)
```

**Comment by jasonrhodes** [comment]
Thanks for the issue, @danieljuhl -- your point about modifying the prototype directly is a fair one. We're mimicking Bluebird functionality here, so we'll probably stick with `asCallback` for now (and may switch back to Bluebird for performance at some point, if necessary), but we can export our own extension of the Promise class so we don't interfere with the native Promise for anyone who uses our lib.

I think this might work:
```js
module.exports = class CustomPromise extends Promise {
  asCallback() {
    // ...
  }
}
```
I'll get a PR together soon. Thanks again.

**Comment by danieljuhl** [comment]
@jasonrhodes I get that you try to mimick Bluebird. But given the relative small size of the sparkpost client (in relation to a whole project where it is used), wouldn't it make more sense to use fully native Promises (or a custom Promise type, if asCallback is really needed, though it could easily be handled without). Then make it optional for the developers using the client, to change the Promise used internally, as seen with many other libraries/modules. Otherwise Sparkpost would also add another dependency to all projects not using bluebird already.

And if I, as a developer, wants better perf, I could simply change to Bluebird, or another Promise library that is already used.

**Comment by danieljuhl** [comment]
@jasonrhodes further to this.. As I also noted in my comment, `asCallback` is throwing in my system.

I'm using new relic monitoring, and they are wrapping Promises, to catch the errors. This conflicts with the Sparkpost promise implementation, and thus throwing errors in my code on ALL sparkpost requests.

Could you please reconsider my initial thoughts about using a wrapper function instead of changing the Promise.

**Comment by jasonrhodes** [comment]
>Could you please reconsider my initial thoughts about using a wrapper function instead of changing the Promise.

I'm not sure what you mean by this. The proposed solution extends Promise so it wouldn't change the native Promise at that point.

**Comment by jasonrhodes** [comment]
As for what we do with regard to Bluebird, we haven't made a decision on that yet and you make good points that we're considering re: adding another dependency, etc. For now we'll focus on leaving the native Promise prototype alone, but in the future we plan on figuring the Bluebird/performance stuff out in a way that works for the most people. Thanks!

**Comment by danieljuhl** [comment]
@jasonrhodes All sounds good then. I'm more than willing to make a PR, but you mention ES6 code (class and extends), but I don't see any use of classes in node-sparkpost. I assume we should stick with ES5?

**Comment by danieljuhl** [comment]
Unfortunately `node` checks the constructor for Promises, thus extending isn't that easy.

Are you sure you can't accept the concept in #200 ?

**Comment by jasonrhodes** [comment]
@danieljuhl thanks for the PR, we're looking it over now. What do you mean about 'node checks the constructor for Promises', I'm just curious?

**Comment by danieljuhl** [comment]
@jasonrhodes I tried creating a SparkPostPromise inheriting from Promise, but when executed node throws: "TypeError: #<SparkPostPromise> is not a promise". I read that some other people had encountered the same, trying to subClass native Promise. It might be a mistake somewhere in my code, but it made me stop, and go back to the other solution.

**Comment by aydrian** [comment]
This has been released in [v2.1.1](https://github.com/SparkPost/node-sparkpost/releases/tag/v2.1.1). Thank you.

**Comment by jasonrhodes** [comment]
Thanks again, @danieljuhl !
